### ISSUE TITLE
Train t5chem_MT

### ISSUE THREAD

**Issue by XYxiyang** [HelloJocelynLu/t5chem#25]
Thanks for the elegant work!

I'm wondering how to train the model supporting multi-tasks using the USPTO_500_MT dataset, since T5ChemTasks does not contain a type for multi-tasks. Can you provide some tutorials or scripts, or have I missed anything?

Best,

**Comment by HelloJocelynLu** [comment]
Hi,

Sorry for the late reply as somehow I didn't get notified for this new issue.
The model architecture constrains which tasks can be trained together:

* Tasks requiring a classification head can only be trained with other classification tasks
* Tasks requiring a regression head can only be trained with other regression tasks
* Tasks requiring a seq2seq head can only be trained with other seq2seq tasks

For example, you can train together:

* Forward reaction prediction + retrosynthesis + reagent prediction (all seq2seq)
* Reaction yield prediction + property prediction (all regression)

When preparing your data for multi-task training:

* Prepend a customized prompt (prefix) to your inputs to distinguish between tasks
* Use "mixed" as the task_type parameter when training

Example training command:
```
t5chem train --data_dir data/USPTO_500_MT/mixed/ --output_dir model/ --task_type mixed --pretrain models/pretrain/simple/ --num_epoch 30  
```
For reference, examine the format of the files in your downloaded data/USPTO_500_MT/mixed/ directory to see examples of properly formatted multi-task data.

Please let me know if you have other questions.

**Comment by XYxiyang** [comment]
Hi,

Thanks for your clarifications! The results turned out to be very helpful.
However, I noticed a minor issue that: when I apply the single-step retrosynthesis results (predicted reactants of target molecules) to predict products, the products are very different from the origin target molecules. Can we fine-tune t5chem MT so that we can constrain this bidirectional relationship (i.e., the generated reactants can produce identical products)?

Best,

**Comment by HelloJocelynLu** [comment]
Hi,

> the products are very different from the origin target molecules

Yes, the metric you're referring to is commonly called "recovery/success rate" in the literature, with typical values around 70-80% success rate in some published papers if i remembered correctly.

Before proceeding with architecture modifications, I'd like to confirm: Have you successfully reproduced any of the published benchmark results? This would establish a solid baseline for comparison.

> Can we fine-tune t5chem MT so that we can constrain this bidirectional relationship (i.e., the generated reactants can produce identical products)?

**Current Architecture Limitations**

The current model architecture—like most existing approaches—doesn't explicitly enforce the bidirectional relationship between forward and reverse reactions. This is an interesting limitation to address.

## Proposed Solution

If you want to enforce this bidirectional relationship, I suggest:

1. **Two-Step Training Pipeline**:
   - First step: Train for retrosynthesis (backward prediction)
   - Second step: Train for forward reaction prediction
   - Calculate the loss after both steps complete, comparing final outputs with original inputs

2. **Loss Function Considerations**:
   - This approach might reduce performance for individual steps (single-step retrosynthesis or forward prediction)
   - A combined loss function would be beneficial: `total_loss = α·loss_retro + β·loss_forward + γ·loss_cycle`
   - The weights (α, β, γ) become hyperparameters you can tune for optimal performance

This is a conceptual approach I haven't implemented myself, but it should provide a framework for capturing the bidirectional relationship.

Please let me know if you have questions or would like me to clarify any aspect of this proposal.

**Comment by XYxiyang** [comment]
Hi,
Thank you very much for your rapid response! 
> Yes, the metric you're referring to is commonly called "recovery/success rate" in the literature, with typical values around 70-80% success rate in some published papers if i remembered correctly.

I'm particularly interested in this aspect. The closest related metric I found is called "round-trip accuracy", where Chemformer and Retroformer show strong performance. Thanks for your insignts!

> Before proceeding with architecture modifications, I'd like to confirm: Have you successfully reproduced any of the published benchmark results? This would establish a solid baseline for comparison.

Based on your instructurions, I ran the code
```
t5chem train --data_dir data/USPTO_500_MT/mixed/ --output_dir model/ --task_type mixed --pretrain models/pretrain/simple/ --num_epoch 30  
```
and here is the result:
USPTO_500_MT/Products
Top-1: 72.0% || Invalid 3.12%
Top-2: 79.0% || Invalid 7.60%
Top-3: 82.4% || Invalid 10.66%
Top-4: 84.5% || Invalid 12.84%
Top-5: 86.0% || Invalid 14.55%

USPTO_500_MT/Reactants
Top-1: 14.2% || Invalid 10.44%
Top-2: 19.9% || Invalid 11.82%
Top-3: 23.2% || Invalid 13.20%
Top-4: 25.7% || Invalid 14.47%
Top-5: 27.5% || Invalid 15.64%
Which seems not reach the performance in the benchmark. Did I execute the codes unappropirately?

Best,

**Comment by XYxiyang** [comment]
Problem solved, it is caused by not adding `--prefix` in run_prediction.py. The ideal prefix should also contain ":" as well, e.g., `Product:`.

Many thanks to the authors!
